File: src/dev/danzou/kotlin/Day3.kt
package dev.danzou.kotlin

import dev.danzou.kotlin.utils.AdventTestRunner

internal class Day3 : AdventTestRunner() {
    fun priorityOf(char: Char): Int =
        if (char.isUpperCase()) char - 'A' + 27
        else char - 'a' + 1

    override fun part1(input: String): Number {
        return input.split("\n")
            .map { it.toList() }
            .map { Pair(it.subList(0, it.size / 2), it.subList(it.size / 2, it.size)) }
            .map { it.first.intersect(it.second) }
            .map { priorityOf(it.first()) }
            .sum()
    }

    override fun part2(input: String): Number {
        return input.split("\n")
            .map { it.toList() }
            .chunked(3)
            .map { it.reduce(Iterable<Char>::intersect) }
            .map { priorityOf(it.first()) }
            .sum()
    }
}

File: src/main/scala/dev/danzou/scala/Day3.scala
package dev.danzou.scala

import dev.danzou.advent22.AdventTestRunner22

class Day3 extends AdventTestRunner22 {
  private def priorityOf(char: Char): Int = {
    if (char.isUpper) char - 'A' + 27
    else char - 'a' + 1
  }

  override def part1(input: String): Number =
    input.split("\n")
      .map(str => str.toList)
      .map(list =>
        list.slice(0, list.size / 2).intersect(list.slice(list.size / 2, list.size)))
      .map(list => priorityOf(list.head))
      .sum

  override def part2(input: String): Number = ???
}
